### PR TITLE
Notify for new device

### PR DESCRIPTION
**Pull request by damencho**
Adds notifications shown when a new device (not already selected as preferred) is plugged. Giving an option to the user to use it.
Screenshot 2019-05-02 at 18 19 34
Screenshot 2019-05-02 at 18 19 28

### CODE REVIEW

File: modules/devices/mediaDeviceHelper.js [hunk 1]
```diff
@@ -25,6 +25,17 @@ function getNewAudioOutputDevice(newDevices) {
             d.deviceId === selectedAudioOutputDeviceId)) {
         return 'default';
     }
+
+    const settings = APP.store.getState()['features/base/settings'];
```
Comment (hristoterezov) [inline modules/devices/mediaDeviceHelper.js]: Shouldn't we include this logic to the logic above somehow?

Comment (damencho) [reply modules/devices/mediaDeviceHelper.js]: Nope, so the logic above is, if the current output device is not 'default' and is not found in the available devices, return 'default', this is in case of removal.

But the newly added logic is if the above is not true (most probably that device is available in the list of devices, so current ouput was not removed), so we check whether our preferred device is available and if yes return it, otherwise keep the logic as before and return undefined.

So you can divide the logic in two: 
- removal, which is the first one 
- addition, the second one.

File: conference.js [hunk 2]
```diff
@@ -2336,6 +2338,102 @@ export default {
...
+    /**
+     * Finds a new device by comparing new and old array of devices and dispatches
+     * notification with the new device.
+     * @param {MediaDeviceInfo[]} newDevices - The array of new devices we received.
+     * @param {MediaDeviceInfo[]} oldDevices - The array of the old devices we have.
+     * @private
+     */
+    _checkAndNotifyForNewDevice(newDevices, oldDevices) {
+        // let's find the first device that exists only in new devices
+        const newDevice = newDevices.find(
```
Comment (hristoterezov) [inline conference.js]: - I think this will find the first element that is new. What happens if there are more than one new devices? Is this possible?

 - Isn't there another solution with better performance? Are you worried about the performance?

 - I'm wondering if you can use `Array.prototype.includes` or `indexOf`. Maybe the code will be more readable?

File: conference.js [hunk 3]
```diff
@@ -2336,6 +2338,102 @@ export default {
...
+            case 'audiooutput': {
+                deviceTypeKey = 'settings.selectAudioOutput';
+                break;
+            }
+            }
+
+            APP.store.dispatch(showNotification({
+                description,
+                titleKey: 'notify.newDeviceTitle',
+                titleArguments: { 'deviceType': APP.translation.translate(deviceTypeKey).toLowerCase() },
```
Comment (hristoterezov) [inline conference.js]: Is this going to work if we change the language after the notification is displayed?

I vaguely remember there were some i18n tricks to put some special string in the translation text and it will auto translate it. Maybe if something like this exists it will look better in the code. Unfortunately I don't remember...

File: conference.js [hunk 4]
```diff
@@ -2336,6 +2338,102 @@ export default {
...
     },
 
+    /**
+     * Set a device to be currently used, selected by the user.
+     *
+     * @param {MediaDeviceInfo} device - The device to save.
+     * @returns {boolean} - Returns true in order notifications to be dismissed.
+     * @private
+     */
+    _useDevice(device) {
```
Comment (hristoterezov) [inline conference.js]: I think this function can be easily moved somewhere in the react folder, isn't it?

File: conference.js [hunk 5]
```diff
@@ -2336,6 +2338,102 @@ export default {
...
+    },
+
+    /**
+     * Finds a new device by comparing new and old array of devices and dispatches
+     * notification with the new device.
+     * @param {MediaDeviceInfo[]} newDevices - The array of new devices we received.
+     * @param {MediaDeviceInfo[]} oldDevices - The array of the old devices we have.
+     * @private
+     */
+    _checkAndNotifyForNewDevice(newDevices, oldDevices) {
```
Comment (hristoterezov) [inline conference.js]: I think this function can be easily moved somewhere in the react folder, isn't it?

File: conference.js (hunk 2 above)
Comment (damencho) [reply conference.js]: Yep, I see that this is triggered per new device. Ok, I will change it so we handle all new devices in the array ... it may happen different implementation on different browsers ... 

Initially, I tried using both includes and indexOf, but objects are different instances and are not found in that comparison, that's why I did this implementation with two finds so I can compare IDs.
When I plug my camera these are two devices, mic and camera, and I suspect nobody has thousands of devices so we can worry about performance in this case. These are two loops with 5-10 elements.

File: modules/devices/mediaDeviceHelper.js (hunk 1 above)
Comment (hristoterezov) [reply modules/devices/mediaDeviceHelper.js]: What I mean is, if you remove the current output device shouldn't we try to see if the preferred one is available and choose the preferred one instead of the default? Or this shouldn't be possible?

File: conference.js (hunk 3 above)
Comment (damencho) [reply conference.js]: No this will not work, but it is not a big deal as nobody will leave those notifications to stay forever and it is a small text camera, microphone, audio output that will not be translated ... and I'm not sure that generated html elements will be correctly displayed when you insert them in a translation string in a atlaskit notification ....

File: conference.js (hunk 4 above)
Comment (damencho) [reply conference.js]: Well, this code should move to react when we refactor all device handling and remove it from conference.js. I can move it now as it is, like a utility function somewhere, by adding a new param passing dispatch function, but is not worth it at the moment and will be used only by non react code ... so its better to be moved when we do the refactor.

File: conference.js (hunk 5 above)
Comment (damencho) [reply conference.js]: Same as above ^

File: modules/devices/mediaDeviceHelper.js (hunk 1 above)
Comment (damencho) [reply modules/devices/mediaDeviceHelper.js]: That doesn't make sense. If you have a preferred available and you had removed one you should had removed the preferred, otherwise there is a bug in another place.

File: conference.js (hunk 2 above)
Comment (damencho) [reply conference.js]: I just pushed a change for that.

File: conference.js (hunk 4 above)
Comment (hristoterezov) [reply conference.js]: The thing is, if we are only relying on the moment when we are going to refactor conference.js and we continue to put new code there, it makes it harder and harder to refactor later. 

I would agree to leave it here if moving it would involve some significant effort, time and refactoring. But those are new functions and new code. You can just copy paste them into the react folder and everything will continue working as it is working now. 

I don't see a reason why they need to be methods of the conference object since we are not using `this`.

I would also agree with your argument about keeping the device handling on a single place but I don't think this is the current situation anyway. Actually the device handling is all over the place - in conference.js, mediaDeviceHelper.js (which should also be moved into the react folder very soon) and in react/features/devices. For example we already use `setAudioInputDevice` and other functions from the react folder in conference.js.

Comment (damencho) [reply conference.js]: So you want it those to be moved to react/features/devices? 

I don't care at all where I will move it, cause those are currently used only in conference.js only and they do not belong to react folder and once we refactor it, those functions will most probably be deleted.

File: conference.js (hunk 3 above)
Comment (hristoterezov) [reply conference.js]: I'm wondering if this is going to do the job - https://www.i18next.com/translation-function/nesting#passing-nesting-to-interpolated . Would you try it?

If we don't have another way to make it work, why don't you have 3 translation keys for every type (only the type will be different in the text)? Isn't that going to fix the problem?

Comment (damencho) [reply conference.js]: This doesn't work, I had tried it. We need to upgrade i18next to make it work.

File: modules/translation/translation.js [hunk 6]
```diff
@@ -55,13 +55,6 @@ class Translation {
         // XXX i18next expects undefined if options are missing.
         selector.localize(options ? options : undefined);
     }
-
-    /**
-     * Translates a given key.
-     */
-    translate(key: string, options: Object) {
```
Comment (paweldomas) [inline modules/translation/translation.js]: why is this method being removed in a totally unrelated commit ?

Comment (paweldomas) [reply modules/translation/translation.js]: oh it was it's last usage
